### PR TITLE
add batcher config to splunkhec exporter

### LINKED ISSUE #32545: splunkhec exporter: Allows users to enable exporter batching via config
### Component(s)

exporter/splunkhec

### Is your feature request related to a problem? Please describe.

With exporter batching introduced in https://github.com/open-telemetry/opentelemetry-collector/pull/9738, users should be able to leverage this for splunk hec exporter so they can enable batching in the exporter side.

### Describe the solution you'd like

Adds exporter batching config in the splunk hec exporter config, if the batching is enabled, add the batching options when creating a new log/metrics instance.

```
type Config struct {
   // ...
   BatcherConfig exporterbatcher.Config    `mapstructure:"batcher"`
}

func createLogsExporter() {
 var options []exporterhelper.Option

  cfg := config.(*Config)
  if cfg.BatcherConfig.Enabled {
      options = append(options, exporterhelper.WithBatcher())
  }
  
  return exporterhelper.NewLogsExporter(..., options...)
}
```

### Describe alternatives you've considered

N/A. Users can utilitize batcher processor but the functionality introduce an async gap in the pipeline. This is what the exporter batching can eliminate.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by splunkericl**
**Description:** 
Add a new feature in splunk hec exporter
- utilitize batching framework https://github.com/open-telemetry/opentelemetry-collector/pull/9738 in splunk hec exporter
- adds batcher config in splunk hec exporter config and append it as exporter option if enabled

**Link to tracking Issue:**
Resolves https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32545

**Testing:**
Added unit test for the config

**Documentation:**
Updated the README file

### CODE REVIEW

**Comment by crobert-1** [comment]
Hey @splunkericl, hope you don't mind, I updated your [PR description to automatically close the issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#description-guidelines). 👍

File: .chloggen/exporter-splunkhec-addbatcherconfig.yaml
Comment (crobert-1) [inline .chloggen/exporter-splunkhec-addbatcherconfig.yaml:10]:
```suggestion
note: add exporter batcher config
```
Nit: In the final release notes the component name will prefix this note, so the component will be obvious from that.

File: exporter/splunkhecexporter/README.md
Comment (crobert-1) [inline exporter/splunkhecexporter/README.md:69]: Adding a link here is super helpful, but the link doesn't seem to include any information for the [actual config option](https://github.com/open-telemetry/opentelemetry-collector/blob/7b046d9aaf94e92b2bfc6c25bec4ac33ba5082f2/exporter/exporterbatcher/config.go#L15). I don't actually see any reference outside of code, so I don't have a good answer for what to have here.

Also, the comment above the config gives me a bit of hesitation with adopting the exporter batcher at this point. @dmitryax, do you think it's worth moving forward now, or should we wait to adopt it? Especially considering the Splunk HEC exporter is considered beta.

Comment (splunkericl) [reply exporter/splunkhecexporter/README.md:69]: yeah the reference is just a placeholder when the documentation eventually surfaces. By default the batcher config is disabled so it shouldn't affect existing users. We can:
- remove this from README reference if we don't want users to try it or
- add experimental tags in the config

Having batching enabled in splunk hec exporter is in our short term roadmap so we are hoping there is at least an option to enable it.

Comment (crobert-1) [reply exporter/splunkhecexporter/README.md:69]: That all makes sense, thanks for giving more information here! I'll defer to the code owners then. 👍

Comment (dmitryax) [reply exporter/splunkhecexporter/README.md:69]: Let's put an experimental warning here 👍

Comment (splunkericl) [reply exporter/splunkhecexporter/README.md:69]: I updated the documentations around. Let me know what you think @crobert-1 @dmitryax

File: exporter/splunkhecexporter/config_test.go
Comment (crobert-1) [inline exporter/splunkhecexporter/config_test.go:94]: Is there a test that ensures the batching functionality is disabled by default?

File: exporter/splunkhecexporter/README.md
Comment (crobert-1) [inline exporter/splunkhecexporter/README.md:69]:
```suggestion
- `batcher`(Experimental, disabled by default): Specifies batching configuration on the exporter. This configuration is experimental and may change until the [issue](https://github.com/open-telemetry/opentelemetry-collector/issues/8122) is resolved. Information about the configuration can be found [here](https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterhelper/README.md)
```
Nit: You're welcome to re-word however you see fit, I just thought it may be helpful to specify that this is disabled by default.

File: .chloggen/exporter-splunkhec-addbatcherconfig.yaml
Comment (crobert-1) [inline .chloggen/exporter-splunkhec-addbatcherconfig.yaml:10]:
```suggestion
note: add experimental exporter batcher config
```
Since the component is already specified we probably don't need to include again here. 👍

File: exporter/splunkhecexporter/config_test.go
Comment (splunkericl) [reply exporter/splunkhecexporter/config_test.go:94]: It is tested as part of base test case(defaultCfg). Do you think that is sufficient?

File: exporter/splunkhecexporter/README.md
Comment (crobert-1) [inline exporter/splunkhecexporter/README.md:69]:
```suggestion
- `batcher`(Experimental, disabled by default): Specifies batching configuration on the exporter. Information about the configuration can be found [here](https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterhelper/README.md).
```
I think we should probably remove the issue reference, because the batcher configuration can still change after the issue is closed, and people reading a README probably don't need that kind of specific information. I think references to how to use the configuration, and that it's `Experimental` is good enough, especially since you linked to the issue in the config 👍

File: exporter/splunkhecexporter/config.go
Comment (crobert-1) [inline exporter/splunkhecexporter/config.go:74]: Would it make sense to include an example in the README of how to use the batcher config? I realize it's changing so we may end up getting a stale config, so I'll leave it up to you 👍

**Review by crobert-1 — APPROVED** [review]

File: exporter/splunkhecexporter/config.go
Comment (splunkericl) [reply exporter/splunkhecexporter/config.go:74]: yeah since it is experimental maybe we can leave it out for now. I think at some point there should be a centralized reference in exporter helper with examples. Since every exporter(that are interested in batching) can point to that.

File: exporter/splunkhecexporter/README.md
Comment (splunkericl) [reply exporter/splunkhecexporter/README.md:69]: fixed

File: exporter/splunkhecexporter/factory.go
Comment (dmitryax) [inline exporter/splunkhecexporter/factory.go:71]: This condition is not supposed to be needed. Can be dropped after https://github.com/open-telemetry/opentelemetry-collector/pull/10076 is merged

**Review by atoulme — APPROVED** [review]

File: exporter/splunkhecexporter/factory.go
Comment (atoulme) [reply exporter/splunkhecexporter/factory.go:71]: approved that change.

**Comment by splunkericl** [comment]
waiting for https://github.com/open-telemetry/opentelemetry-collector/pull/10076 to merge in so the exporter doesn't need the enabled if statement check

**Comment by dmitryax** [comment]
@splunkericl the PR in core is merged and pulled to this repo. Please rebase and remove the condition

**Comment by splunkericl** [comment]
Thanks @dmitryax . i have rebased and remove the if-statement.

File: exporter/splunkhecexporter/factory.go
Comment (dmitryax) [inline exporter/splunkhecexporter/factory.go:94]: Let's use `exporterbatcher.NewDefaultConfig()`, but disabled
```
batcherCfg := exporterbatcher.NewDefaultConfig()
batcherCfg.Enabled = false
```

File: exporter/splunkhecexporter/factory.go
Comment (dmitryax) [inline exporter/splunkhecexporter/factory.go:61]: This extraction is unrelated now. Better just to add `exporterhelper.WithBatcher(cfg.BatcherConfig),` lines separately to keep the changelog clean

File: exporter/splunkhecexporter/factory.go
Comment (dmitryax) [reply exporter/splunkhecexporter/factory.go:94]: Because `MinSizeConfig` has to be provided, otherwise it's noop

Comment (splunkericl) [reply exporter/splunkhecexporter/factory.go]: Fixed the initialization.

File: exporter/splunkhecexporter/factory.go
Comment (splunkericl) [reply exporter/splunkhecexporter/factory.go]: Sure. put back optional to inline

**Review by dmitryax — APPROVED** [review]

**Pull request merged**
